File: frontend-sdk/src/types/productivity.ts
export type FocusAgentNoteStatus = "active" | "archived";
export type FocusAgentTaskStatus = "todo" | "in_progress" | "completed" | "archived";
export type FocusAgentProductivitySourceKind = "chat" | "agent_team" | "merge_review" | "task_output" | "artifact" | "manual" | string;

export interface FocusAgentNote {
  note_id: string;
  user_id: string;
  title: string;
  body: string;
  tags: string[];
  status: FocusAgentNoteStatus;
  source_thread_id?: string | null;
  source_artifact_id?: string | null;
  source_kind?: FocusAgentProductivitySourceKind | null;
  source_id?: string | null;
  source_url?: string | null;
  pinned_context?: string | null;
  captured_from?: string | null;
  is_archived: boolean;
  metadata: Record<string, unknown>;
  created_at: string;
  updated_at: string;
  archived_at?: string | null;
}

export interface FocusAgentTask {
  task_id: string;
  user_id: string;
  title: string;
  description: string;
  status: FocusAgentTaskStatus;
  due_at?: string | null;
  priority?: number | null;
  source_thread_id?: string | null;
  source_note_id?: string | null;
  source_kind?: FocusAgentProductivitySourceKind | null;
  source_id?: string | null;
  source_url?: string | null;
  pinned_context?: string | null;
  captured_from?: string | null;
  assignee_user_id?: string | null;
  tags: string[];
  metadata: Record<string, unknown>;
  created_at: string;
  updated_at: string;
  completed_at?: string | null;
  archived_at?: string | null;
}

export interface FocusAgentTaskEvent {
  event_id: string;
  task_id: string;
  user_id: string;
  kind: "created" | "updated" | "completed" | "archived";
  data: Record<string, unknown>;
  created_at: string;
}

export interface FocusAgentCreateNoteRequest {
  title: string;
  body?: string;
  tags?: string[];
  source_thread_id?: string | null;
  source_artifact_id?: string | null;
  source_kind?: FocusAgentProductivitySourceKind | null;
  source_id?: string | null;
  source_url?: string | null;
  pinned_context?: string | null;
  captured_from?: string | null;
  metadata?: Record<string, unknown> | null;
}

export interface FocusAgentUpdateNoteRequest {
  title?: string | null;
  body?: string | null;
  tags?: string[] | null;
  status?: FocusAgentNoteStatus | null;
  source_thread_id?: string | null;
  source_artifact_id?: string | null;
  source_kind?: FocusAgentProductivitySourceKind | null;
  source_id?: string | null;
  source_url?: string | null;
  pinned_context?: string | null;
  captured_from?: string | null;
  is_archived?: boolean | null;
  metadata?: Record<string, unknown> | null;
}

export interface FocusAgentListNotesRequest {
  q?: string | null;
  tag?: string[] | null;
  source_kind?: FocusAgentProductivitySourceKind | null;
  include_archived?: boolean;
  limit?: number;
  offset?: number;
}

export interface FocusAgentNoteResponse {
  note: FocusAgentNote;
}

export interface FocusAgentNoteListResponse {
  items: FocusAgentNote[];
  count: number;
}

export interface FocusAgentCreateTaskRequest {
  title: string;
  description?: string;
  due_at?: string | null;
  priority?: number | null;
  source_thread_id?: string | null;
  source_note_id?: string | null;
  source_kind?: FocusAgentProductivitySourceKind | null;
  source_id?: string | null;
  source_url?: string | null;
  pinned_context?: string | null;
  captured_from?: string | null;
  assignee_user_id?: string | null;
  tags?: string[];
  metadata?: Record<string, unknown> | null;
}

export interface FocusAgentUpdateTaskRequest {
  title?: string | null;
  description?: string | null;
  status?: FocusAgentTaskStatus | null;
  due_at?: string | null;
  priority?: number | null;
  source_thread_id?: string | null;
  source_note_id?: string | null;
  source_kind?: FocusAgentProductivitySourceKind | null;
  source_id?: string | null;
  source_url?: string | null;
  pinned_context?: string | null;
  captured_from?: string | null;
  assignee_user_id?: string | null;
  tags?: string[] | null;
  metadata?: Record<string, unknown> | null;
}

export interface FocusAgentListTasksRequest {
  status?: FocusAgentTaskStatus | null;
  source_kind?: FocusAgentProductivitySourceKind | null;
  include_archived?: boolean;
  limit?: number;
  offset?: number;
}

export interface FocusAgentTaskResponse {
  task: FocusAgentTask;
}

export interface FocusAgentTaskListResponse {
  items: FocusAgentTask[];
  count: number;
}

export interface FocusAgentTaskEventListResponse {
  items: FocusAgentTaskEvent[];
  count: number;
}

export interface FocusAgentCaptureNoteRequest {
  title?: string | null;
  body: string;
  tags?: string[];
  source_kind?: FocusAgentProductivitySourceKind | null;
  source_id?: string | null;
  source_url?: string | null;
  source_thread_id?: string | null;
  source_artifact_id?: string | null;
  pinned_context?: string | null;
  captured_from?: string | null;
  metadata?: Record<string, unknown> | null;
}

export interface FocusAgentCaptureTaskRequest {
  title: string;
  description?: string;
  due_at?: string | null;
  priority?: number | null;
  tags?: string[];
  source_kind?: FocusAgentProductivitySourceKind | null;
  source_id?: string | null;
  source_url?: string | null;
  source_thread_id?: string | null;
  source_note_id?: string | null;
  pinned_context?: string | null;
  captured_from?: string | null;
  assignee_user_id?: string | null;
  metadata?: Record<string, unknown> | null;
}

File: src/focus_agent/core/productivity.py
from __future__ import annotations

from enum import StrEnum
from typing import Any

from pydantic import BaseModel, Field, model_validator


class FocusTaskStatus(StrEnum):
    TODO = "todo"
    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"
    ARCHIVED = "archived"


class FocusNoteStatus(StrEnum):
    ACTIVE = "active"
    ARCHIVED = "archived"


class FocusTaskEventKind(StrEnum):
    CREATED = "created"
    UPDATED = "updated"
    COMPLETED = "completed"
    ARCHIVED = "archived"


class ProductivityCaptureKind(StrEnum):
    NOTE = "note"
    TASK = "task"


class FocusNote(BaseModel):
    note_id: str
    user_id: str
    title: str
    body: str = ""
    tags: list[str] = Field(default_factory=list)
    status: FocusNoteStatus = FocusNoteStatus.ACTIVE
    source_thread_id: str | None = None
    source_artifact_id: str | None = None
    source_kind: str | None = None
    source_id: str | None = None
    source_url: str | None = None
    pinned_context: dict[str, Any] = Field(default_factory=dict)
    captured_from: str | None = None
    is_archived: bool = False
    metadata: dict[str, Any] = Field(default_factory=dict)
    created_at: str
    updated_at: str
    archived_at: str | None = None

    @model_validator(mode="after")
    def _sync_archive_fields(self) -> FocusNote:
        if self.is_archived and self.status != FocusNoteStatus.ARCHIVED:
            self.status = FocusNoteStatus.ARCHIVED
        elif self.status == FocusNoteStatus.ARCHIVED and not self.is_archived:
            self.is_archived = True
        return self


class FocusTask(BaseModel):
    task_id: str
    user_id: str
    title: str
    description: str = ""
    status: FocusTaskStatus = FocusTaskStatus.TODO
    due_at: str | None = None
    priority: int | None = None
    source_thread_id: str | None = None
    source_note_id: str | None = None
    source_artifact_id: str | None = None
    source_kind: str | None = None
    source_id: str | None = None
    source_url: str | None = None
    pinned_context: dict[str, Any] = Field(default_factory=dict)
    captured_from: str | None = None
    assignee_user_id: str | None = None
    tags: list[str] = Field(default_factory=list)
    metadata: dict[str, Any] = Field(default_factory=dict)
    created_at: str
    updated_at: str
    completed_at: str | None = None
    archived_at: str | None = None


class FocusTaskEvent(BaseModel):
    event_id: str
    task_id: str
    user_id: str
    kind: FocusTaskEventKind
    data: dict[str, Any] = Field(default_factory=dict)
    created_at: str


__all__ = [
    "FocusNote",
    "FocusNoteStatus",
    "FocusTask",
    "FocusTaskEvent",
    "FocusTaskEventKind",
    "FocusTaskStatus",
    "ProductivityCaptureKind",
]
